{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "e6e5a00e-36fe-4489-ae69-33c53c84e759",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "source": [
    "# Tables\n",
    "This notebook contains the code showing how the tables in the paper is generated from the data. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "8d22dd33-0163-4382-952f-c208af21eabb",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-----\n",
      "numpy               1.26.4\n",
      "pandas              1.5.3\n",
      "session_info        1.0.0\n",
      "-----\n",
      "IPython             8.18.1\n",
      "jupyter_client      8.6.3\n",
      "jupyter_core        5.8.1\n",
      "jupyterlab          4.4.5\n",
      "notebook            7.4.5\n",
      "-----\n",
      "Python 3.9.5 | packaged by conda-forge | (default, Jun 19 2021, 00:27:35) [Clang 11.1.0 ]\n",
      "macOS-15.6.1-x86_64-i386-64bit\n",
      "-----\n",
      "Session information updated at 2025-09-19 10:17\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "import session_info\n",
    "\n",
    "session_info.show(html=False)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "3facdbb8-4bd4-4d0a-ae2f-6855466b4e09",
   "metadata": {},
   "source": [
    "## Table 1.\n",
    "Distribution of  HLA and Tri-SNP haplotypes among TEDDY samples.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "47fffe79-3d4c-4af7-b055-1a9cde2a7dcf",
   "metadata": {},
   "source": [
    "### Import metadata\n",
    "Metadata for table 1 contains all of the samples where HLA type and tri-SNP genotyping was available (n=7759)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "61d6a78a-3bb8-437e-b25d-23d67ad071d0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mp257_maskid</th>\n",
       "      <th>hla_type</th>\n",
       "      <th>tri</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>996512</td>\n",
       "      <td>DR4/DR4</td>\n",
       "      <td>010/010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>474979</td>\n",
       "      <td>DR4/DR8</td>\n",
       "      <td>000/010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>581412</td>\n",
       "      <td>DR4/DR4</td>\n",
       "      <td>010/010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>864158</td>\n",
       "      <td>DR3/DR4</td>\n",
       "      <td>010/101</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>669152</td>\n",
       "      <td>DR3/DR4</td>\n",
       "      <td>010/101</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   mp257_maskid hla_type      tri\n",
       "0        996512  DR4/DR4  010/010\n",
       "1        474979  DR4/DR8  000/010\n",
       "2        581412  DR4/DR4  010/010\n",
       "3        864158  DR3/DR4  010/101\n",
       "4        669152  DR3/DR4  010/101"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meta = pd.read_csv(\"../data/meta_hla.csv\")\n",
    "meta.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9ddc6da1-1374-44c1-a7de-b655d4355ac5",
   "metadata": {},
   "source": [
    "### HLA type distribution among all samples"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "438d2d8c-b37a-4ae0-9aeb-cd5918d3c4e7",
   "metadata": {},
   "source": [
    "Convert HLA type and tri-SNP genotype columns to ordered categorical for consistency in table order."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "912fbb45-9e15-41ed-b0c3-2647e18e04da",
   "metadata": {},
   "outputs": [],
   "source": [
    "hla_order = ['DR3/DR3', 'DR3/DR4', 'DR4/DR4', 'DR4/DR8', 'DR1/DR4',\n",
    "             'DR4/DR13', 'DR3/DR9', 'DR4/DR9', 'DR4/DR4*030X/020X',\n",
    "             'DR4/DR4*030X/0304']\n",
    "\n",
    "meta[\"hla_type\"] = pd.Categorical(\n",
    "    meta[\"hla_type\"], categories=hla_order, ordered=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "58fbabbc-529c-489a-ad72-97b3f3acc21b",
   "metadata": {},
   "outputs": [],
   "source": [
    "tri_order = ['010/010', '010/101','101/101',\n",
    "             '000/010', '001/010', '000/101',\n",
    "             '000/000', '000/001', '001/101',\n",
    "             '011/101', '010/011']\n",
    "\n",
    "meta[\"tri\"] = pd.Categorical(\n",
    "    meta[\"tri\"], categories=tri_order,  ordered=True)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4869f7bd-0675-4689-afa8-8ab52771e31b",
   "metadata": {},
   "source": [
    "Create the disrtibution table. This is main text Table 1."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "82399051-2716-485b-ac17-5c221bae4318",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>tri</th>\n",
       "      <th>010/010</th>\n",
       "      <th>010/101</th>\n",
       "      <th>101/101</th>\n",
       "      <th>000/010</th>\n",
       "      <th>001/010</th>\n",
       "      <th>000/101</th>\n",
       "      <th>000/000</th>\n",
       "      <th>000/001</th>\n",
       "      <th>001/101</th>\n",
       "      <th>011/101</th>\n",
       "      <th>010/011</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hla_type</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>DR3/DR3</th>\n",
       "      <td>62</td>\n",
       "      <td>348</td>\n",
       "      <td>1177</td>\n",
       "      <td>13</td>\n",
       "      <td>2</td>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DR3/DR4</th>\n",
       "      <td>408</td>\n",
       "      <td>2550</td>\n",
       "      <td>3</td>\n",
       "      <td>19</td>\n",
       "      <td>8</td>\n",
       "      <td>30</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DR4/DR4</th>\n",
       "      <td>1480</td>\n",
       "      <td>5</td>\n",
       "      <td>3</td>\n",
       "      <td>39</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DR4/DR8</th>\n",
       "      <td>14</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1297</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>17</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DR1/DR4</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>15</td>\n",
       "      <td>143</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DR4/DR13</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>56</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DR3/DR9</th>\n",
       "      <td>2</td>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DR4/DR9</th>\n",
       "      <td>13</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DR4/DR4*030X/020X</th>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DR4/DR4*030X/0304</th>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "tri                010/010  010/101  101/101  000/010  001/010  000/101  \\\n",
       "hla_type                                                                  \n",
       "DR3/DR3                 62      348     1177       13        2       11   \n",
       "DR3/DR4                408     2550        3       19        8       30   \n",
       "DR4/DR4               1480        5        3       39        3        1   \n",
       "DR4/DR8                 14        0        0     1297        5        0   \n",
       "DR1/DR4                  0        0        0       15      143        0   \n",
       "DR4/DR13                 0        0        0        0       56        0   \n",
       "DR3/DR9                  2       16        0        0        0        0   \n",
       "DR4/DR9                 13        1        0        0        0        0   \n",
       "DR4/DR4*030X/020X        4        0        0        0        0        0   \n",
       "DR4/DR4*030X/0304        3        0        0        0        0        0   \n",
       "\n",
       "tri                000/000  000/001  001/101  011/101  010/011  \n",
       "hla_type                                                        \n",
       "DR3/DR3                  0        0        2        1        0  \n",
       "DR3/DR4                  0        1        1        1        1  \n",
       "DR4/DR4                  0        0        0        0        0  \n",
       "DR4/DR8                 17        1        0        0        0  \n",
       "DR1/DR4                  0        2        0        0        0  \n",
       "DR4/DR13                 0        1        0        0        0  \n",
       "DR3/DR9                  0        0        0        0        0  \n",
       "DR4/DR9                  0        0        0        0        0  \n",
       "DR4/DR4*030X/020X        0        0        0        0        0  \n",
       "DR4/DR4*030X/0304        0        0        0        0        0  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dist = meta.pivot_table(\n",
    "    index=\"hla_type\", columns=\"tri\", values=\"mp257_maskid\",\n",
    "    aggfunc=\"count\").fillna(0).astype(int)\n",
    "dist"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a64359ea-152a-4e47-a1c7-d9128c64e198",
   "metadata": {},
   "source": [
    "Show values each tri-SNP genotype as percentage of whole. This is the \"All\" row in table 1 of the published paper."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "8b40e929-c5f3-4e6c-9dd5-26f68a9986e5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "tri\n",
       "010/010    1986 (25.6%)\n",
       "010/101    2920 (37.6%)\n",
       "101/101    1183 (15.2%)\n",
       "000/010    1383 (17.8%)\n",
       "001/010      217 (2.8%)\n",
       "000/101       42 (0.5%)\n",
       "000/000       17 (0.2%)\n",
       "000/001        5 (0.1%)\n",
       "001/101        3 (0.0%)\n",
       "011/101        2 (0.0%)\n",
       "010/011        1 (0.0%)\n",
       "dtype: object"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sum_row = dist.sum().astype(str) + \" (\" + (round(dist.sum().div(dist.sum().sum()) * 100, 1)).astype(str) + \"%)\"\n",
    "sum_row"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "97a9c312-f8c8-4a55-bec3-8cb4fbb5bec9",
   "metadata": {},
   "source": [
    "Show values each HLA type as percentage of whole. This is the \"All\" column of Table 1 in the published paper."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "2b8f148c-66c2-4ed5-942d-facba3ed4bc1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "hla_type\n",
       "DR3/DR3              1616 (20.8%)\n",
       "DR3/DR4              3022 (38.9%)\n",
       "DR4/DR4              1531 (19.7%)\n",
       "DR4/DR8              1334 (17.2%)\n",
       "DR1/DR4                160 (2.1%)\n",
       "DR4/DR13                57 (0.7%)\n",
       "DR3/DR9                 18 (0.2%)\n",
       "DR4/DR9                 14 (0.2%)\n",
       "DR4/DR4*030X/020X        4 (0.1%)\n",
       "DR4/DR4*030X/0304        3 (0.0%)\n",
       "dtype: object"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sum_col = dist.sum(axis=1).astype(str) + \" (\" + (\n",
    "    round(dist.sum(axis=1).div(dist.sum(axis=1).sum(), axis=0) * 100, 1)).astype(str) + \"%)\"\n",
    "sum_col"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "58e5af2a-5979-4e57-9af3-fd87e9762c46",
   "metadata": {},
   "source": [
    "## Supplementary Tables 1-6\n",
    "Descriptive characteristics of children with respect to tested outcomes."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "40e60926-dc67-4df4-bf66-5688cf1abc21",
   "metadata": {},
   "source": [
    "### Import meta data\n",
    "These tables are created from the \"filtered\" final analysis set that contains 7703 children. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "071110c2-0b14-44ed-a2bc-aa31a59cadd2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mp257_maskid</th>\n",
       "      <th>family_id</th>\n",
       "      <th>immunochip_id</th>\n",
       "      <th>sex</th>\n",
       "      <th>Cntry</th>\n",
       "      <th>POP</th>\n",
       "      <th>ancestry</th>\n",
       "      <th>POP_reported</th>\n",
       "      <th>t1d</th>\n",
       "      <th>t1d_diag_age_censor</th>\n",
       "      <th>...</th>\n",
       "      <th>rs61751041 (LAMB1)</th>\n",
       "      <th>rs6967298 (AUTS2)</th>\n",
       "      <th>rs72704176 (ASH1L)</th>\n",
       "      <th>rs72717025 (FCGR2A)</th>\n",
       "      <th>rs73043122 (RNASET2/MIR3939)</th>\n",
       "      <th>rs77532435 (GRB10)</th>\n",
       "      <th>rs8013918 (FOS)</th>\n",
       "      <th>rs9934817 (RBFOX1)</th>\n",
       "      <th>rs389884</th>\n",
       "      <th>rs926552</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>996512</td>\n",
       "      <td>9000247.0</td>\n",
       "      <td>9000247_996512</td>\n",
       "      <td>Male</td>\n",
       "      <td>SWE</td>\n",
       "      <td>EUR</td>\n",
       "      <td>EUR</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>135.972</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>474979</td>\n",
       "      <td>474979.0</td>\n",
       "      <td>474979_474979</td>\n",
       "      <td>Male</td>\n",
       "      <td>SWE</td>\n",
       "      <td>EUR</td>\n",
       "      <td>EUR</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>183.443</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>581412</td>\n",
       "      <td>581412.0</td>\n",
       "      <td>581412_581412</td>\n",
       "      <td>Male</td>\n",
       "      <td>SWE</td>\n",
       "      <td>EUR</td>\n",
       "      <td>EUR</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>186.235</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>864158</td>\n",
       "      <td>864158.0</td>\n",
       "      <td>864158_864158</td>\n",
       "      <td>Male</td>\n",
       "      <td>SWE</td>\n",
       "      <td>EUR</td>\n",
       "      <td>EUR</td>\n",
       "      <td>EUR</td>\n",
       "      <td>0</td>\n",
       "      <td>167.017</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>669152</td>\n",
       "      <td>669152.0</td>\n",
       "      <td>669152_669152</td>\n",
       "      <td>Male</td>\n",
       "      <td>SWE</td>\n",
       "      <td>EUR</td>\n",
       "      <td>EUR</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>174.080</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 61 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   mp257_maskid  family_id   immunochip_id   sex Cntry  POP ancestry  \\\n",
       "0        996512  9000247.0  9000247_996512  Male   SWE  EUR      EUR   \n",
       "1        474979   474979.0   474979_474979  Male   SWE  EUR      EUR   \n",
       "2        581412   581412.0   581412_581412  Male   SWE  EUR      EUR   \n",
       "3        864158   864158.0   864158_864158  Male   SWE  EUR      EUR   \n",
       "4        669152   669152.0   669152_669152  Male   SWE  EUR      EUR   \n",
       "\n",
       "  POP_reported  t1d  t1d_diag_age_censor  ...  rs61751041 (LAMB1)  \\\n",
       "0          NaN    1              135.972  ...                 0.0   \n",
       "1          NaN    0              183.443  ...                 1.0   \n",
       "2          NaN    0              186.235  ...                 0.0   \n",
       "3          EUR    0              167.017  ...                 0.0   \n",
       "4          NaN    0              174.080  ...                 0.0   \n",
       "\n",
       "   rs6967298 (AUTS2)  rs72704176 (ASH1L)  rs72717025 (FCGR2A)  \\\n",
       "0                0.0                 0.0                  0.0   \n",
       "1                0.0                 0.0                  0.0   \n",
       "2                0.0                 0.0                  0.0   \n",
       "3                0.0                 0.0                  0.0   \n",
       "4                1.0                 0.0                  0.0   \n",
       "\n",
       "   rs73043122 (RNASET2/MIR3939)  rs77532435 (GRB10)  rs8013918 (FOS)  \\\n",
       "0                           0.0                 0.0              1.0   \n",
       "1                           0.0                 0.0              2.0   \n",
       "2                           0.0                 0.0              0.0   \n",
       "3                           0.0                 0.0              1.0   \n",
       "4                           1.0                 0.0              2.0   \n",
       "\n",
       "   rs9934817 (RBFOX1) rs389884  rs926552  \n",
       "0                 0.0      0.0       0.0  \n",
       "1                 0.0      0.0       0.0  \n",
       "2                 0.0      0.0       0.0  \n",
       "3                 0.0      1.0       1.0  \n",
       "4                 0.0      1.0       0.0  \n",
       "\n",
       "[5 rows x 61 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filtered = pd.read_csv(\"../data/filtered_meta_final.csv\")\n",
    "filtered.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8ae4e3d6-9371-4805-be76-5b274d9cf3a4",
   "metadata": {},
   "source": [
    "Convert the used covariates to ordered categoricals for consistency."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "d8cb40f6-4952-4551-bd90-e0961164b5e1",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "select_pops = [\"EUR\", \"AMR\", \"AFR\"]\n",
    "filtered[\"POP\"] = pd.Categorical(\n",
    "    filtered[\"POP\"], categories=select_pops, ordered=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "2a544a29-8ef8-4d4f-9742-cc4b8190c06a",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "ctry = [\"US\", \"SWE\", \"FIN\", \"GER\"]\n",
    "filtered[\"Country\"] = pd.Categorical(\n",
    "    filtered[\"Cntry\"], categories=ctry, ordered=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6a2deef4-ff01-46f3-847e-5598b186a65d",
   "metadata": {},
   "source": [
    "Change column names for the tested outcomes to match those reported in the paper."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "5bba09a3-da35-475c-825e-0a01f978ec96",
   "metadata": {},
   "outputs": [],
   "source": [
    "filtered.rename(\n",
    "    columns={\"t1d\": \"T1D\", \"persist_conf_ab\": \"IA\",\n",
    "             \"celiac_diagnosis\": \"CD\", \"tga_persist\": \"CDA\"},\n",
    "    inplace=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "15c6c554-0809-4e0c-b5d1-fc6faaa5d1c2",
   "metadata": {},
   "source": [
    "Add an additional column which is same for all rows to be used in total number rows/columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "a7c2abd2-8a8d-4ae5-8635-9a0b72ce0ed6",
   "metadata": {},
   "outputs": [],
   "source": [
    "filtered[\"All\"] = \"All\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3629570b-7b1e-4be7-9b4d-f07729339f1b",
   "metadata": {},
   "source": [
    "Create a list of covariates and outcomes for 6 tables."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "f46102c1-77e6-4a45-96f0-8a42c9a11c79",
   "metadata": {},
   "outputs": [],
   "source": [
    "covariates = [\"sex\", \"POP\", \"hla_type\", \"Country\", \"FDR\", \"All\"]\n",
    "outcomes = [\"IA\", \"T1D\", \"GADA first\", \"IAA first\", \"CD\", \"CDA\"]          "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "06e4edad-4913-49d2-80ed-87f305672580",
   "metadata": {},
   "source": [
    "Loop through outcomes and covariates, print supplementary tables 1-6."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "6da595fb-4c9b-4eaa-8ba1-4315db56d12f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Supplementary Table 1.\n",
      "\n",
      "IA                          No         Yes           All\n",
      "sex      Female    3388 (49.5)  390 (45.5)   3778 (49.0)\n",
      "         Male      3457 (50.5)  468 (54.5)   3925 (51.0)\n",
      "POP      EUR       6122 (89.4)  796 (92.8)   6918 (89.8)\n",
      "         AMR         647 (9.5)    56 (6.5)     703 (9.1)\n",
      "         AFR          76 (1.1)     6 (0.7)      82 (1.1)\n",
      "hla_type DR1/DR4     142 (2.1)    18 (2.1)     160 (2.1)\n",
      "         DR3/DR3   1486 (21.7)  121 (14.1)   1607 (20.9)\n",
      "         DR3/DR4   2599 (38.0)  418 (48.7)   3017 (39.2)\n",
      "         DR4/DR13     47 (0.7)    10 (1.2)      57 (0.7)\n",
      "         DR4/DR4   1371 (20.0)  158 (18.4)   1529 (19.8)\n",
      "         DR4/DR8   1200 (17.5)  133 (15.5)   1333 (17.3)\n",
      "Country  US        2885 (42.1)  297 (34.6)   3182 (41.3)\n",
      "         SWE       2019 (29.5)  290 (33.8)   2309 (30.0)\n",
      "         FIN       1482 (21.7)  214 (24.9)   1696 (22.0)\n",
      "         GER         459 (6.7)    57 (6.6)     516 (6.7)\n",
      "FDR      No        6155 (89.9)  691 (80.5)   6846 (88.9)\n",
      "         Yes        690 (10.1)  167 (19.5)    857 (11.1)\n",
      "All      All       6845 (88.9)  858 (11.1)  7703 (100.0)\n",
      "\n",
      "\n",
      "Supplementary Table 2.\n",
      "\n",
      "T1D                         No         Yes           All\n",
      "sex      Female    3597 (49.2)  181 (45.6)   3778 (49.0)\n",
      "         Male      3709 (50.8)  216 (54.4)   3925 (51.0)\n",
      "POP      EUR       6552 (89.7)  366 (92.2)   6918 (89.8)\n",
      "         AMR         674 (9.2)    29 (7.3)     703 (9.1)\n",
      "         AFR          80 (1.1)     2 (0.5)      82 (1.1)\n",
      "hla_type DR1/DR4     146 (2.0)    14 (3.5)     160 (2.1)\n",
      "         DR3/DR3   1571 (21.5)    36 (9.1)   1607 (20.9)\n",
      "         DR3/DR4   2797 (38.3)  220 (55.4)   3017 (39.2)\n",
      "         DR4/DR13     51 (0.7)     6 (1.5)      57 (0.7)\n",
      "         DR4/DR4   1457 (19.9)   72 (18.1)   1529 (19.8)\n",
      "         DR4/DR8   1284 (17.6)   49 (12.3)   1333 (17.3)\n",
      "Country  US        3034 (41.5)  148 (37.3)   3182 (41.3)\n",
      "         SWE       2204 (30.2)  105 (26.4)   2309 (30.0)\n",
      "         FIN       1591 (21.8)  105 (26.4)   1696 (22.0)\n",
      "         GER         477 (6.5)    39 (9.8)     516 (6.7)\n",
      "FDR      No        6560 (89.8)  286 (72.0)   6846 (88.9)\n",
      "         Yes        746 (10.2)  111 (28.0)    857 (11.1)\n",
      "All      All       7306 (94.8)   397 (5.2)  7703 (100.0)\n",
      "\n",
      "\n",
      "Supplementary Table 3.\n",
      "\n",
      "GADA first                  No         Yes           All\n",
      "sex      Female    3601 (49.2)  177 (46.3)   3778 (49.0)\n",
      "         Male      3720 (50.8)  205 (53.7)   3925 (51.0)\n",
      "POP      EUR       6565 (89.7)  353 (92.4)   6918 (89.8)\n",
      "         AMR         678 (9.3)    25 (6.5)     703 (9.1)\n",
      "         AFR          78 (1.1)     4 (1.0)      82 (1.1)\n",
      "hla_type DR1/DR4     157 (2.1)     3 (0.8)     160 (2.1)\n",
      "         DR3/DR3   1522 (20.8)   85 (22.3)   1607 (20.9)\n",
      "         DR3/DR4   2828 (38.6)  189 (49.5)   3017 (39.2)\n",
      "         DR4/DR13     55 (0.8)     2 (0.5)      57 (0.7)\n",
      "         DR4/DR4   1472 (20.1)   57 (14.9)   1529 (19.8)\n",
      "         DR4/DR8   1287 (17.6)   46 (12.0)   1333 (17.3)\n",
      "Country  US        3036 (41.5)  146 (38.2)   3182 (41.3)\n",
      "         SWE       2164 (29.6)  145 (38.0)   2309 (30.0)\n",
      "         FIN       1622 (22.2)   74 (19.4)   1696 (22.0)\n",
      "         GER         499 (6.8)    17 (4.5)     516 (6.7)\n",
      "FDR      No        6531 (89.2)  315 (82.5)   6846 (88.9)\n",
      "         Yes        790 (10.8)   67 (17.5)    857 (11.1)\n",
      "All      All       7321 (95.0)   382 (5.0)  7703 (100.0)\n",
      "\n",
      "\n",
      "Supplementary Table 4.\n",
      "\n",
      "IAA first                   No         Yes           All\n",
      "sex      Female    3636 (49.2)  142 (45.4)   3778 (49.0)\n",
      "         Male      3754 (50.8)  171 (54.6)   3925 (51.0)\n",
      "POP      EUR       6624 (89.6)  294 (93.9)   6918 (89.8)\n",
      "         AMR         685 (9.3)    18 (5.8)     703 (9.1)\n",
      "         AFR          81 (1.1)     1 (0.3)      82 (1.1)\n",
      "hla_type DR1/DR4     151 (2.0)     9 (2.9)     160 (2.1)\n",
      "         DR3/DR3   1578 (21.4)    29 (9.3)   1607 (20.9)\n",
      "         DR3/DR4   2870 (38.8)  147 (47.0)   3017 (39.2)\n",
      "         DR4/DR13     52 (0.7)     5 (1.6)      57 (0.7)\n",
      "         DR4/DR4   1471 (19.9)   58 (18.5)   1529 (19.8)\n",
      "         DR4/DR8   1268 (17.2)   65 (20.8)   1333 (17.3)\n",
      "Country  US        3082 (41.7)  100 (31.9)   3182 (41.3)\n",
      "         SWE       2215 (30.0)   94 (30.0)   2309 (30.0)\n",
      "         FIN       1597 (21.6)   99 (31.6)   1696 (22.0)\n",
      "         GER         496 (6.7)    20 (6.4)     516 (6.7)\n",
      "FDR      No        6600 (89.3)  246 (78.6)   6846 (88.9)\n",
      "         Yes        790 (10.7)   67 (21.4)    857 (11.1)\n",
      "All      All       7390 (95.9)   313 (4.1)  7703 (100.0)\n",
      "\n",
      "\n",
      "Supplementary Table 5.\n",
      "\n",
      "CD                          No         Yes           All\n",
      "sex      Female    3413 (48.2)  365 (59.2)   3778 (49.0)\n",
      "         Male      3673 (51.8)  252 (40.8)   3925 (51.0)\n",
      "POP      EUR       6325 (89.3)  593 (96.1)   6918 (89.8)\n",
      "         AMR         681 (9.6)    22 (3.6)     703 (9.1)\n",
      "         AFR          80 (1.1)     2 (0.3)      82 (1.1)\n",
      "hla_type DR1/DR4     155 (2.2)     5 (0.8)     160 (2.1)\n",
      "         DR3/DR3   1308 (18.5)  299 (48.5)   1607 (20.9)\n",
      "         DR3/DR4   2805 (39.6)  212 (34.4)   3017 (39.2)\n",
      "         DR4/DR13     54 (0.8)     3 (0.5)      57 (0.7)\n",
      "         DR4/DR4   1444 (20.4)   85 (13.8)   1529 (19.8)\n",
      "         DR4/DR8   1320 (18.6)    13 (2.1)   1333 (17.3)\n",
      "Country  US        2955 (41.7)  227 (36.8)   3182 (41.3)\n",
      "         SWE       2049 (28.9)  260 (42.1)   2309 (30.0)\n",
      "         FIN       1594 (22.5)  102 (16.5)   1696 (22.0)\n",
      "         GER         488 (6.9)    28 (4.5)     516 (6.7)\n",
      "FDR      No        6447 (96.4)  496 (80.7)   6943 (95.1)\n",
      "         Yes         242 (3.6)  119 (19.3)     361 (4.9)\n",
      "All      All       7086 (92.0)   617 (8.0)  7703 (100.0)\n",
      "\n",
      "\n",
      "Supplementary Table 6.\n",
      "\n",
      "CDA                         No          Yes           All\n",
      "sex      Female    2555 (47.2)   735 (56.8)   3290 (49.0)\n",
      "         Male      2860 (52.8)   559 (43.2)   3419 (51.0)\n",
      "POP      EUR       4855 (89.7)  1237 (95.6)   6092 (90.8)\n",
      "         AMR         515 (9.5)     55 (4.3)     570 (8.5)\n",
      "         AFR          45 (0.8)      2 (0.2)      47 (0.7)\n",
      "hla_type DR1/DR4     134 (2.5)      8 (0.6)     142 (2.1)\n",
      "         DR3/DR3    870 (16.1)   530 (41.0)   1400 (20.9)\n",
      "         DR3/DR4   2146 (39.6)   503 (38.9)   2649 (39.5)\n",
      "         DR4/DR13     46 (0.8)      4 (0.3)      50 (0.7)\n",
      "         DR4/DR4   1130 (20.9)   193 (14.9)   1323 (19.7)\n",
      "         DR4/DR8   1089 (20.1)     56 (4.3)   1145 (17.1)\n",
      "Country  US        2232 (41.2)   471 (36.4)   2703 (40.3)\n",
      "         SWE       1588 (29.3)   479 (37.0)   2067 (30.8)\n",
      "         FIN       1262 (23.3)   272 (21.0)   1534 (22.9)\n",
      "         GER         333 (6.1)     72 (5.6)     405 (6.0)\n",
      "FDR      No        5118 (96.5)  1119 (86.9)   6237 (94.6)\n",
      "         Yes         186 (3.5)   169 (13.1)     355 (5.4)\n",
      "All      All       5415 (80.7)  1294 (19.3)  6709 (100.0)\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for i in range(len(outcomes)):\n",
    "    o = outcomes[i]\n",
    "    tab_list = []\n",
    "    # set the correct FDR column for Celiac vs T1D\n",
    "    if o.startswith(\"CD\"):\n",
    "        filtered[\"FDR\"] = filtered[\"celiac_fdr\"].replace(\n",
    "            {0: \"No\", 1: \"Yes\"})\n",
    "    else:\n",
    "        filtered[\"FDR\"] = filtered[\"FDR-T1D\"].replace(\n",
    "            {0: \"No\", 1: \"Yes\"})\n",
    "\n",
    "    # loop through covariates and create a pivot table for each\n",
    "    for cov in covariates:\n",
    "        tab = filtered.pivot_table(\n",
    "            index=cov, columns=o,\n",
    "            values=\"mp257_maskid\", aggfunc=\"count\")\n",
    "        tab.rename(columns={0: \"No\", 1: \"Yes\"}, inplace=True)\n",
    "\n",
    "        # create \"All\" column in the pivot table for total counts\n",
    "        tab_sum = tab.sum(axis=1)\n",
    "        tab[\"All\"] = tab_sum\n",
    "\n",
    "        # Determine total numbers to use for calculating percentage\n",
    "        # values. These will be different for actual covariates and\n",
    "        # the pseudo-covariate \"All\".\n",
    "        # We use the \"All\" column from the main data as a covariate\n",
    "        # for the totals row. This is different from the \"All\" column\n",
    "        # in the pivot table created above.\n",
    "        if cov == \"All\":\n",
    "            total = tab_sum\n",
    "            axis = 0\n",
    "        else:\n",
    "            total = tab.sum()\n",
    "            axis = 1\n",
    "\n",
    "        # calculate values as percentage and add within parantheses\n",
    "        tab_pct = tab.astype(str) + \" (\" + (\n",
    "            round(tab.div(total, axis=axis) * 100, 1)).astype(str) + \")\"\n",
    "        tab_list.append(tab_pct)\n",
    "    # concatenate tables from all covariates for this outcome\n",
    "    outcome_tab = pd.concat(tab_list, keys=covariates)\n",
    "\n",
    "    # print the table\n",
    "    print(\"Supplementary Table {}.\".format(i+1))\n",
    "    print()\n",
    "    print(outcome_tab)\n",
    "    print()\n",
    "    print()\n",
    "\n",
    "\n",
    "  "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
